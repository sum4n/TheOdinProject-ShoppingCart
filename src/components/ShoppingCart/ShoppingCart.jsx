import { Link, useOutletContext } from "react-router-dom";
import styles from "./ShoppingCart.module.css";

const ShoppingCart = () => {
  const { itemsInCart, addToCart, deleteFromCart } = useOutletContext();

  // console.log(itemsInCart);

  let total = 0;
  if (itemsInCart.length != 0) {
    for (let i = 0; i < itemsInCart.length; i++) {
      total += itemsInCart[i].price * itemsInCart[i].quantity;
    }
  }

  function inputHandler(item, quantity) {
    if (Number(quantity) && parseInt(quantity) > 0 && parseInt(quantity) < 11) {
      addToCart(item, quantity);
    } else {
      alert("Invalid input");
    }
  }

  function handleIncrement(item, quantity) {
    if (parseInt(quantity) < 10) {
      addToCart(item, quantity + 1);
    } else {
      alert("Can not add more than 10 of the same item.");
    }
  }
  function handleDecrement(item, quantity) {
    if (parseInt(quantity) > 1) {
      addToCart(item, quantity - 1);
    }
  }

  function handleDeleteFromCart(item) {
    if (confirm("Do you want to delete this item?")) {
      deleteFromCart(item);
    }
  }

  return (
    <div className={styles.cartBody}>
      <h1>Shopping Cart</h1>
      <ol>
        {itemsInCart.map((item) => {
          return (
            <div key={item.id}>
              <hr />
              <li className={styles.itemList}>
                <img className={styles.itemImage} src={item.image} alt="" />
                <div>
                  <Link
                    to={`/products/${item.id}`}
                    className={styles.itemTitle}
                  >
                    {item.title}
                  </Link>
                  <p>Price: {item.price}$</p>
                  <div>
                    <label className={styles.inputLabel} htmlFor="quantity">
                      Quantity
                    </label>
                    <input
                      className={styles.inputBox}
                      type="number"
                      id="quantity"
                      name="quantity"
                      max={10}
                      min={1}
                      // defaultValue={1}
                      value={item.quantity}
                      onChange={(event) =>
                        inputHandler(item, event.target.value)
                      }
                    />
                    <button
                      onClick={() => handleIncrement(item, item.quantity)}
                      className={styles.btnQuantity}
                      // disabled={item.quantity}
                    >
                      +
                    </button>
                    {item.quantity > 1 ? (
                      <button
                        onClick={() => handleDecrement(item, item.quantity)}
                        className={styles.btnQuantity}
                      >
                        -
                      </button>
                    ) : (
                      <button
                        onClick={() => handleDeleteFromCart(item)}
                        className={styles.btnDelete}
                      >
                        Remove from cart
                      </button>
                    )}
                  </div>
                  <p>Total price: {item.quantity * item.price}$</p>
                </div>
              </li>
            </div>
          );
        })}
        <hr />
      </ol>
      <p>Total amount: {Math.round(total * 100) / 100}$</p>
      <button className={styles.btnCheckout}>Checkout</button>
    </div>
  );
};

export default ShoppingCart;
