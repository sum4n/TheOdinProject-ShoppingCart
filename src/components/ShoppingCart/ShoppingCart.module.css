.cartBody {
  min-width: 350px;
  padding: 0px 5px 5px 10px;
  border-radius: 6px;
  box-shadow: 0px 0px 8px rgb(63, 63, 63);
  margin: 0 2% 0 2%;
}

.btnCheckout {
  font-weight: normal;
}

.itemList {
  display: flex;
  gap: 24px;
  align-items: center;
}

.itemImage {
  width: 100px;
  height: 100px;
}

.itemTitle {
  font-weight: bold;
  text-decoration: underline;
}

.quantityBox {
  /* border: 1px solid black; */
  display: flex;
  align-items: center;
}

.btnQuantity,
.btnDelete {
  margin-left: 14px;
  font-size: 18px;
  /* height: 50px; */
  min-width: 140px;
  max-width: auto;
}

.btnDelete {
  font-size: 14px;
  width: 140px;
  /* font-weight: normal; */
}
