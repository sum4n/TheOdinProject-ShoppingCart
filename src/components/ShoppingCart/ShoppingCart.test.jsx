import { render, screen } from "@testing-library/react";
import { afterEach, beforeEach, describe, expect, vi } from "vitest";
import ShoppingCart from "./ShoppingCart";
import { MemoryRouter, Outlet, Route, Routes } from "react-router-dom";

const itemsInCart = [
  {
    id: 1,
    title: "Product A Title",
    description: "Product A Description",
    image: "Product A img URL",
    price: 5,
    quantity: 1,
  },
  {
    id: 2,
    title: "Product B Title",
    description: "Product B description",
    image: "Product B img URL",
    price: 10,
    quantity: 3,
  },
];

const addToCart = vi.fn();
const deleteFromCart = vi.fn();

let renderResult;

describe("ShoppingCart component", () => {
  beforeEach(() => {
    renderResult = render(
      <MemoryRouter initialEntries={["/cart"]}>
        <Routes>
          <Route
            element={
              <Outlet context={{ itemsInCart, addToCart, deleteFromCart }} />
            }
          >
            <Route path="/cart" element={<ShoppingCart />} />
          </Route>
        </Routes>
      </MemoryRouter>
    );
  });

  it("Renders exactly 2 items", () => {
    const items = screen.getAllByTestId("cart-item");
    expect(items).toHaveLength(2);
  });

  it("Renders total amount correctly", () => {
    let totalPrice = 0;
    itemsInCart.forEach((item) => {
      totalPrice += item.price * item.quantity;
    });

    expect(
      screen.getByText(`Total amount: ${totalPrice}$`)
    ).toBeInTheDocument();
  });

  it("Matches snapshot", () => {
    expect(renderResult.container).toMatchSnapshot();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });
});
