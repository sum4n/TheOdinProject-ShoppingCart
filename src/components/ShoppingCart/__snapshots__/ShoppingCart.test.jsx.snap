// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShoppingCart component > Matches snapshot 1`] = `
<div>
  <div
    class="_cartBody_b6b9fb"
  >
    <h1>
      Shopping Cart
    </h1>
    <ol>
      <div
        data-testid="cart-item"
      >
        <hr />
        <li
          class="_itemList_b6b9fb"
        >
          <img
            alt="Product A Title"
            class="_itemImage_b6b9fb"
            src="Product A img URL"
          />
          <div>
            <a
              class="_itemTitle_b6b9fb"
              href="/products/1"
            >
              Product A Title
            </a>
            <p>
              Price: 
              5
              $
            </p>
            <div
              class="_quantityBox_b6b9fb"
            >
              <label
                class="_inputLabel_b6b9fb"
                for="quantity"
              >
                Quantity
              </label>
              <input
                class="_inputBox_b6b9fb"
                id="quantity"
                max="10"
                min="1"
                name="quantity"
                type="number"
                value="1"
              />
              <button
                class="_btnQuantity_b6b9fb"
              >
                +
              </button>
              <button
                class="_btnDelete_b6b9fb"
              >
                Remove from cart
              </button>
            </div>
            <p>
              Total price: 
              5
              $
            </p>
          </div>
        </li>
      </div>
      <div
        data-testid="cart-item"
      >
        <hr />
        <li
          class="_itemList_b6b9fb"
        >
          <img
            alt="Product B Title"
            class="_itemImage_b6b9fb"
            src="Product B img URL"
          />
          <div>
            <a
              class="_itemTitle_b6b9fb"
              href="/products/2"
            >
              Product B Title
            </a>
            <p>
              Price: 
              10
              $
            </p>
            <div
              class="_quantityBox_b6b9fb"
            >
              <label
                class="_inputLabel_b6b9fb"
                for="quantity"
              >
                Quantity
              </label>
              <input
                class="_inputBox_b6b9fb"
                id="quantity"
                max="10"
                min="1"
                name="quantity"
                type="number"
                value="3"
              />
              <button
                class="_btnQuantity_b6b9fb"
              >
                +
              </button>
              <button
                class="_btnQuantity_b6b9fb"
              >
                -
              </button>
            </div>
            <p>
              Total price: 
              30
              $
            </p>
          </div>
        </li>
      </div>
      <hr />
    </ol>
    <p>
      Total amount: 
      35
      $
    </p>
    <button
      class="_btnCheckout_b6b9fb"
    >
      Checkout
    </button>
  </div>
</div>
`;
